commit 400928b83bd254d20f0790d467f221bb52707e6e
Author: Tom Lane <tgl@sss.pgh.pa.us>
Date:   Mon Jan 29 12:06:07 2024 -0500

    Fix incompatibilities with libxml2 >= 2.12.0.
    
    libxml2 changed the required signature of error handler callbacks
    to make the passed xmlError struct "const".  This is causing build
    failures on buildfarm member caiman, and no doubt will start showing
    up in the field quite soon.  Add a version check to adjust the
    declaration of xml_errorHandler() according to LIBXML_VERSION.
    
    2.12.x also produces deprecation warnings for contrib/xml2/xpath.c's
    assignment to xmlLoadExtDtdDefaultValue.  I see no good reason for
    that to still be there, seeing that we disabled external DTDs (at a
    lower level) years ago for security reasons.  Let's just remove it.
    
    Back-patch to all supported branches, since they might all get built
    with newer libxml2 once it gets a bit more popular.  (The back
    branches produce another deprecation warning about xpath.c's use of
    xmlSubstituteEntitiesDefault().  We ought to consider whether to
    back-patch all or part of commit 65c5864d7 to silence that.  It's
    less urgent though, since it won't break the buildfarm.)
    
    Discussion: https://postgr.es/m/1389505.1706382262@sss.pgh.pa.us

diff --git a/contrib/xml2/xpath.c b/contrib/xml2/xpath.c
index a967257546a..b999b1f7066 100644
--- a/contrib/xml2/xpath.c
+++ b/contrib/xml2/xpath.c
@@ -72,12 +72,10 @@ pgxml_parser_init(PgXmlStrictness strictness)
 	/* Note: we're assuming an elog cannot be thrown by the following calls */
 
 	/* Initialize libxml */
 	xmlInitParser();
 
-	xmlLoadExtDtdDefaultValue = 1;
-
 	return xmlerrcxt;
 }
 
 
 /* Encodes special characters (<, >, &, " and \r) as XML entities */
diff --git a/src/backend/utils/adt/xml.c b/src/backend/utils/adt/xml.c
index f869c680afd..3e24aba546f 100644
--- a/src/backend/utils/adt/xml.c
+++ b/src/backend/utils/adt/xml.c
@@ -65,10 +65,20 @@
  * testing the library version number.
  */
 #if LIBXML_VERSION >= 20704
 #define HAVE_XMLSTRUCTUREDERRORCONTEXT 1
 #endif
+
+/*
+ * libxml2 2.12 decided to insert "const" into the error handler API.
+ */
+#if LIBXML_VERSION >= 21200
+#define PgXmlErrorPtr const xmlError *
+#else
+#define PgXmlErrorPtr xmlErrorPtr
+#endif
+
 #endif							/* USE_LIBXML */
 
 #include "access/htup_details.h"
 #include "access/table.h"
 #include "catalog/namespace.h"
@@ -122,11 +132,11 @@ struct PgXmlErrorContext
 
 static xmlParserInputPtr xmlPgEntityLoader(const char *URL, const char *ID,
 										   xmlParserCtxtPtr ctxt);
 static void xml_errsave(Node *escontext, PgXmlErrorContext *errcxt,
 						int sqlcode, const char *msg);
-static void xml_errorHandler(void *data, xmlErrorPtr error);
+static void xml_errorHandler(void *data, PgXmlErrorPtr error);
 static int	errdetail_for_xml_code(int code);
 static void chopStringInfoNewlines(StringInfo str);
 static void appendStringInfoLineSeparator(StringInfo str);
 
 #ifdef USE_LIBXMLCONTEXT
@@ -2022,11 +2032,11 @@ xml_errsave(Node *escontext, PgXmlErrorContext *errcxt,
 
 /*
  * Error handler for libxml errors and warnings
  */
 static void
-xml_errorHandler(void *data, xmlErrorPtr error)
+xml_errorHandler(void *data, PgXmlErrorPtr error)
 {
 	PgXmlErrorContext *xmlerrcxt = (PgXmlErrorContext *) data;
 	xmlParserCtxtPtr ctxt = (xmlParserCtxtPtr) error->ctxt;
 	xmlParserInputPtr input = (ctxt != NULL) ? ctxt->input : NULL;
 	xmlNodePtr	node = error->node;
